### ISSUE TITLE
race_condition_ttl is not really implemented or wrongly implemented

### ISSUE THREAD

**Issue by zben** [rails/rails#34918]
### Steps to reproduce
This issue is related to the intended purpose/definition of race_condition_ttl on Rails.cache.

### Expected behavior
From https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html, it says

> Setting :race_condition_ttl is very useful in situations where a cache entry is used very frequently and is under heavy load. If a cache expires and due to heavy load several different processes will try to read data natively and then they all will try to write to cache. To avoid that case the first process to find an expired cache entry will bump the cache expiration time by the value set in :race_condition_ttl. Yes, this process is extending the time for a stale value by another few seconds. Because of extended life of the previous cache, other processes will continue to use slightly stale data for a just a bit longer. In the meantime that first process will go ahead and will write into cache the new value. After that all the processes will start getting the new value. The key is to keep :race_condition_ttl small.

### Actual behavior
In the[default redis adapter](https://github.com/rails/rails/blob/cd16413ef3280a50c1b13b1ed7be1a907fad88d1/activesupport/lib/active_support/cache/redis_cache_store.rb#L391), race_condition_ttl value is ignored and it merely adds 5 minutes to the expires_in value, which does not give the above intended behavior. 
[default mem cache adapter](https://github.com/rails/rails/blob/b1223458979adc29117ffa1392526c6a166effed/activesupport/lib/active_support/cache/mem_cache_store.rb#L156) also just adds 5 minutes to expires_in. 
Other cache_stores defined in activesupport does not even mention race_condition_ttl.

I might be missing something, but it seems this feature is either not implemented or implemented naively. If it is equivalent to adding 5 minutes to expires_in, then people can just increase their expires_in without needing this extra option. 

Also it seems that at least for redis, once expiration time is reached, the value will be deleted from redis so if 2 threads try to fetch value at that time after that cache key is expired, there is no old value to extend expiration for. Maybe I am missing something.

In conclusion, the implementation of race_condition_ttl in various default cache stores are either missing or different from intended behavior. Maybe this option should be dropped from Rails or at least add a note in the documentation that it is not fully supported by any cache stores currently and is only a theoretically useful option. 

### System configuration
**Rails version**:
5.0
**Ruby version**:
2.3.1

**Comment by rafaelfranca** [comment]
Did you check this https://github.com/rails/rails/blob/cd16413ef3280a50c1b13b1ed7be1a907fad88d1/activesupport/lib/active_support/cache.rb#L694-L700?

**Comment by zben** [comment]
I missed that part of the code. Now it makes a lot of sense. The 5 minutes extension of expires_in is to allow the option to cache old value for a bit longer. It was not super clear what the 5 minutes was about. Thanks!

**Comment by nroose** [comment]
Seems like a funny name. The problem it is trying to solve is a problem. But not a race condition. And depending on what you set it to, it has the potential to create a race condition. And when the above says set it to small, I that means "longer than it takes to calculate the new value but less than 5 minutes."

**Comment by zzak** [comment]
@nroose Maybe this is an opportunity to improve the documentation. How about a PR?
